### PR TITLE
Fixes & Refactoring Galore

### PR DESCRIPTION
**Pull request by birkelund**
There's a lot of stuff here. All patch build up to introducing `struct iommu_ctx` to generalize the backends.

Within the `43 files changed, 1184 insertions(+), 1083 deletions(-)`, there are some fixes here and there as well.

This is a shit-ton of code to review. My sincere apologies.

### CODE REVIEW

File: src/iommu/context.c [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+#include "context.h"
+
+#ifdef HAVE_VFIO_DEVICE_BIND_IOMMUFD
+static bool __iommufd_broken;
+
+static void __attribute__((constructor)) __check_iommufd_broken(void)
+{
+	struct stat sb;
+
+	if (stat("/dev/vfio/devices", &sb) || (sb.st_mode & S_IFDIR)) {
```
Comment (birkelund) [inline src/iommu/context.c:33]: Well, this is wrong...

File: src/iommu/iommufd.c [hunk 2]
```diff
@@ -168,8 +168,7 @@ int vfio_get_device_fd(struct vfio_container *vfio, const char *bdf)
 
 close_dev:
 	/* closing devfd will automatically unbind it from iommufd */
-	if (close(devfd))
-		log_error("failed to close cdev fd: %d\n", devfd);
+	log_fatal_if(close(devfd), "close\n");
```
Comment (Joelgranados) [inline src/iommu/iommufd.c:153]: Could we give a bit more information on the error? like the original message?

That goes for all the log_fata_if on this patch.

File: src/iommu/vfio.c [hunk 3]
```diff
@@ -213,6 +210,26 @@ static int vfio_configure_iommu(struct vfio_container *vfio)
...
+
+	if (ioctl(group->fd, VFIO_GROUP_SET_CONTAINER, &vfio->fd)) {
+		log_debug("failed to add group to vfio container\n");
+		return -1;
+	}
+
+	if (vfio_configure_iommu(vfio)) {
+		log_error("failed to configure iommu\n");
+
+		log_fatal_if(ioctl(group->fd, VFIO_GROUP_UNSET_CONTAINER), "unset container\n");
```
Comment (Joelgranados) [inline src/iommu/vfio.c:216]: nice!. That answers my initial question that I left on the code.

File: src/iommu/iommufd.c [hunk 4]
```diff
@@ -273,12 +280,12 @@ struct iommu_ctx *get_iommu_ctx(const char *name)
 	return &ioas->ctx;
```
[the remark this answers is not in the document]
Comment (birkelund) [reply src/iommu/iommufd.c:280]: We don't have any destroy functions currently, so the contexts cannot be deallocated. But no, the memory is not leaked. We can always get the `ioas` struct back using `container_of` and we do so internally.

File: src/iommu/vfio.c [hunk 5]
```diff
@@ -436,8 +432,11 @@ struct iommu_ctx *get_iommu_context(const char *name)
 	return &vfio->ctx;
```
[the remark this answers is not in the document]
Comment (birkelund) [reply src/iommu/vfio.c:432]: Nope. See above.

File: src/iommu/context.h [hunk 6]
```diff
@@ -30,4 +30,12 @@ struct iommu_ctx {
 	unsigned int flags;
 };
 
-struct iommu_ctx *get_default_iommu_ctx(void);
+struct iommu_ctx *iommu_get_default_context(void);
+
```
[the remark this answers is not in the document]
Comment (birkelund) [reply src/iommu/context.h:34]: No :) We can switch it back :)

File: include/vfn/iommu/dma.h [hunk 7]
```diff
@@ -40,8 +40,9 @@ enum iommu_map_flags {
  * If @vaddr falls within an already mapped area, calculate the corresponding
  * iova instead.
  *
- * Note that the allocated IOVA can never be reused in the lifetime of the
- * process, so use this for mapping memory that will be reused.
+ * Note that, for the vfio backend, the allocated IOVA can never be reused in
+ * the lifetime of the process, so use this for mapping memory that will be
+ * reused.
```
[the remark this answers is not in the document]
Comment (birkelund) [reply include/vfn/iommu/dma.h:45]: I'll rephrase it. It is to convey the fact that IOVAs (in the vfio backend) are not recycled when unmapping memory. Once used, you can't map any memory to it again.

File: include/vfn/nvme.h [hunk 8]
```diff
@@ -35,6 +35,7 @@ extern "C" {
 #include <vfn/support.h>
 #include <vfn/trace.h>
 #include <vfn/trace/events.h>
+#include <vfn/iommu.h>
 #include <vfn/vfio.h>
 #include <vfn/iommu.h>
```
[the remark this answers is not in the document]
Comment (birkelund) [reply include/vfn/nvme.h:40]: Woops.

File: src/vfio/container.c
[the remark this answers is not in the document]
Comment (birkelund) [reply src/vfio/container.c:1]: Github is being weird. I can't see the context that you refer to.

File: src/iommu/vfio.c [hunk 9]
```diff
@@ -208,7 +208,7 @@ static int vfio_configure_iommu(struct vfio_container *vfio)
 		log_debug("failed to get iommu capabilities\n");
 		return -1;
 	}
-#endif //VFIO_IOMMU_INFO_CAPS
```
[the remark this answers is not in the document]
Comment (birkelund) [reply src/iommu/vfio.c:210]: I'm not sure if checkpatch catches it. I *should* but I also do not think checkpatch was enabled for PRs against next, so it couldn't when your stuff was merged.

File: src/iommu/iommufd.c (hunk 2 above)
Comment (birkelund) [reply src/iommu/iommufd.c:153]: Yeah, true. I will readd that.

File: src/iommu/vfio.c (hunk 3 above)
Comment (birkelund) [reply src/iommu/vfio.c:216]: Yeah, and we want to consider it fatal if we cant unset here. If we couldn't unset, I dont know what the application or library should do to fix it. Something worse must be wrong.

File: src/iommu/context.c [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+#include "context.h"
+
+#ifdef HAVE_VFIO_DEVICE_BIND_IOMMUFD
+static bool __iommufd_broken;
+
+static void __attribute__((constructor)) __check_iommufd_broken(void)
+{
+	struct stat sb;
+
+	if (stat("/dev/vfio/devices", &sb) || !S_ISDIR(sb.st_mode)) {
```
[the remark this answers is not in the document]
Comment (Joelgranados) [reply src/iommu/context.c:33]: Indeed. and searching for the configuration value in /proc does not work in general as the kernel does not always have that available

[the remark this answers is not in the document]
Comment (Joelgranados) [reply src/iommu/context.c:33]: How about we just do this :
https://github.com/Joelgranados/libvfn/commit/c190243b802d53d9ba89af23cdd8047621906bcc

File: src/vfio/container.c
[the remark this answers is not in the document]
Comment (Joelgranados) [reply src/vfio/container.c:1]: Yep, because I commented on a commit. Let me try to find that commit.

[the remark this answers is not in the document]
Comment (Joelgranados) [reply src/vfio/container.c:1]: I believe I was referring to this commit https://github.com/OpenMPDK/libvfn/pull/9/commits/e5b48832bf00de590972f6ad5058a64f22f91ae7

File: src/iommu/vfio.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (Joelgranados) [reply src/iommu/vfio.c:210]: I got some checkpatch messages on one of my original PR versions, so we do run it when a PR is being evaluated. Not sure if it checks for comment formatting.

**Pull request closed without merging**